clients/src/main/java/org/apache/kafka/common/requests/AddPartitionsToTxnRequest.java

```diff
@@ -118,11 +123,78 @@ public AddPartitionsToTxnRequestData data() {
 
     @Override
     public AddPartitionsToTxnResponse getErrorResponse(int throttleTimeMs, Throwable e) {
-        final HashMap<TopicPartition, Errors> errors = new HashMap<>();
-        for (TopicPartition partition : partitions()) {
-            errors.put(partition, Errors.forException(e));
+        Errors error = Errors.forException(e);
+        AddPartitionsToTxnResponseData response = new AddPartitionsToTxnResponseData();
+        if (version < 4) {
+            response.setResultsByTopicV3AndBelow(errorResponseForTopics(data.v3AndBelowTopics(), error));
+        } else {
+            AddPartitionsToTxnResultCollection results = new AddPartitionsToTxnResultCollection();
+            for (AddPartitionsToTxnTransaction transaction : data().transactions()) {
+                results.add(errorResponseForTransaction(transaction.transactionalId(), error));
+            }
+            response.setResultsByTransaction(results);
+            response.setErrorCode(error.code());
```

Comment:
Yeah, I like the top level error code for that reason. However, we must be clear on how we want to use it. I suppose that we can only use it for cluster authorization failure and unexpected errors failing the entire request/response.

Reply:
To clarify -- I'm not sure if being clear on how we want to use it (ie the errors it is used for) is the same question as if we want to include the error on the txns as well. 

As for the latter, I think we are more error prone if we set the error to none. I guess I'm just not sure what we are trying to accomplish by removing them from the lower fields. To be clear, I think I've also seen this pattern on other responses with top level errors.

Reply:
I was looking at the AlterPartition API as a reference. There, when we have a top level error, we don't set the partitions in the response at all. We basically save space. See [here](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/requests/AlterPartitionRequest.java#L46).

Reply:
Don't forget to take the version into account here. We can only do this for v4.

Reply:
Yup. This will only be done in v4 responses. So for now, we don't do any checks since the server side usage is upcoming in the next PR.

As a side note, seems that AlterPartition includes the top level error in errorCounts so I did that here as well.